### PR TITLE
Use Alembic to create db schema

### LINKED ISSUE #36: Setup a database migration tool
As we gear up for production, we need to ensure the database schema is being managed. We need to setup a tool to track DB changes, the ability to spin up a DB from scratch, and seed data if needed.

Acceptance Criteria:
- The tool can create the schema in an empty database
- The tool can seed an empty DB assuming the schema exists
- The tool can track changes made to the schema, and revert changes if needed

Suggestions:
- Use [Alembic](https://alembic.sqlalchemy.org/en/latest/index.html) b/c it's the standard for python development
- Use ChatGPT to guide you through the process if it's your 1st time

References:
- [Alembic Auto-generation](https://alembic.sqlalchemy.org/en/latest/autogenerate.html)
- [Alembic Tutorial](https://alembic.sqlalchemy.org/en/latest/tutorial.html)
- https://www.youtube.com/watch?v=i9RX03zFDHU&t=249s

>In the context of Python databases, Alembic is a lightweight database migration tool for usage with the SQLAlchemy Database Toolkit for Python. Alembic provides a way to manage database schema changes over time, ensuring that the database schema remains in sync with the application's models.

> Key features of Alembic include:

>Version Control: Alembic maintains a history of all the schema changes applied to the database, allowing you to upgrade or downgrade the schema to any point in time.
Script Generation: It can automatically generate migration scripts by comparing the current database schema with the schema defined in SQLAlchemy models.
Command Line Interface: Alembic provides a command-line tool to apply migrations, roll back migrations, and generate new migration scripts.
Integration with SQLAlchemy: Since Alembic is built for SQLAlchemy, it integrates seamlessly with SQLAlchemy's models and sessions.

### PR DESCRIPTION
**Pull request by AshciR**
fixes #36

### CODE REVIEW

File: .env-template [hunk 1]
```diff
@@ -2,4 +2,5 @@ TELEGRAM_API_KEY=
 POSTGRES_USER=
 POSTGRES_PASSWORD=
 POSTGRES_DB=
-POSTGRES_CONTAINER=db
\ No newline at end of file
+POSTGRES_CONTAINER=db
+MIGRATE=true
```
Comment (AshciR) [inline .env-template:6]: Migration will only happen if this is true

File: Dockerfile [hunk 2]
```diff
@@ -18,20 +18,34 @@ ENV PYTHONPATH="/app/src:$PYTHONPATH"
...
 WORKDIR $WORKDIR
 
+# Install necessary tools including netcat
+# We use netcat to check if the db is running prior
+# to running the migration tool
+RUN apt-get update && apt-get install -y netcat-openbsd
+
 # Copy the requirements and src code
-COPY requirements.lock ./
+COPY requirements-dev.lock ./
```
Comment (AshciR) [inline Dockerfile:27]: Changed it b/c alembic is a dev dependency

File: README.md [hunk 3]
```diff
@@ -38,4 +38,26 @@ docker-compose up -d
 Use the following command to build the image
 ```shell
 docker build . -t <tag-the-image>
-```     
\ No newline at end of file
+```
+
+## Running the database
```
Comment (AshciR) [inline README.md:43]: Added instructions on how to use the tool

File: alembic.ini
Comment (AshciR) [inline alembic.ini:1]: This was generated by running `alembic init`. Contains configs for the tool, but we'll likely not touch it

File: alembic/env.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+from shout_subgroup.models import Base
+
+load_dotenv()
+
+POSTGRES_USER = os.getenv('POSTGRES_USER')
+POSTGRES_PASSWORD = os.getenv('POSTGRES_PASSWORD')
+POSTGRES_HOST = os.getenv('POSTGRES_CONTAINER')
+POSTGRES_DB = os.getenv('POSTGRES_DB')
+
+DATABASE_URL = f"postgresql://{POSTGRES_USER}:{POSTGRES_PASSWORD}@{POSTGRES_HOST}:5432/{POSTGRES_DB}"
```
Comment (AshciR) [inline alembic/env.py:18]: Added this. Alembic needs to know which db to apply the changes to

File: alembic/env.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+# this is the Alembic Config object, which provides
+# access to the values within the .ini file in use.
+config = context.config
+config.set_main_option("sqlalchemy.url", DATABASE_URL)
```
Comment (AshciR) [inline alembic/env.py:23]: Replaces the original value in the .ini file

File: alembic/env.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+# Interpret the config file for Python logging.
+# This line sets up loggers basically.
+if config.config_file_name is not None:
+    fileConfig(config.config_file_name)
+
+# add your model's MetaData object here
+# for 'autogenerate' support
+# from myapp import mymodel
+# target_metadata = mymodel.Base.metadata
+target_metadata = Base.metadata
```
Comment (AshciR) [inline alembic/env.py:34]: This is the "magic" that looks at our `models.py` file, then tracks the changes, and creates a schema update if needed when `alembic revision --autogenerate` is ran

File: alembic/script.py.mako
Comment (AshciR) [inline alembic/script.py.mako:1]: Autogenerated during init

File: alembic/versions/fcdf7873db3f_initial_schema.py
Comment (AshciR) [inline alembic/versions/fcdf7873db3f_initial_schema.py:1]: Important file. Whenever we change the db models, a new version will be created; It's similar to git in that sense.
These are committed and during PRs, we look at the upgrade, and downgrade functions. They should align with the expected SQL changes to the DB

File: alembic/versions/fcdf7873db3f_initial_schema.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+
+# revision identifiers, used by Alembic.
+revision: str = 'fcdf7873db3f'
+down_revision: Union[str, None] = None
+branch_labels: Union[str, Sequence[str], None] = None
+depends_on: Union[str, Sequence[str], None] = None
+
+
+def upgrade() -> None:
```
Comment (AshciR) [inline alembic/versions/fcdf7873db3f_initial_schema.py:21]: B/c it's the initial run, it created all the tables the application uses

File: alembic/versions/fcdf7873db3f_initial_schema.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    op.create_table('users_subgroups_join_table',
+    sa.Column('subgroup_id', sa.String(), nullable=True),
+    sa.Column('user_id', sa.String(), nullable=True),
+    sa.ForeignKeyConstraint(['subgroup_id'], ['subgroups.subgroup_id'], ),
+    sa.ForeignKeyConstraint(['user_id'], ['users.user_id'], )
+    )
+    # ### end Alembic commands ###
+
+
+def downgrade() -> None:
```
Comment (AshciR) [inline alembic/versions/fcdf7873db3f_initial_schema.py:69]: If this db version had to be reversed, it'd call this function. In this case, it's simply drops all the tables

File: entrypoint.sh
Comment (AshciR) [inline entrypoint.sh:1]: This is used in `Dockerfile` to run the db migrations if needed

File: pyproject.toml [hunk 9]
```diff
@@ -25,10 +25,20 @@ dev-dependencies = [
...
+    "alembic==1.13.2"
 ]
 
 [tool.hatch.metadata]
 allow-direct-references = true
 
 [tool.hatch.build.targets.wheel]
 packages = ["src/shout_subgroup"]
+
+[tool.rye.scripts]
```
Comment (AshciR) [inline pyproject.toml:37]: Added helper scripts for rye

File: src/shout_subgroup/database.py [hunk 10]
```diff
@@ -16,42 +16,28 @@ def configure_database() -> bool:
     POSTGRES_DB = os.getenv('POSTGRES_DB')
     POSTGRES_CONTAINER = os.getenv('POSTGRES_CONTAINER')
 
+    db_configs = {
+        "POSTGRES_USER": POSTGRES_USER,
+        "POSTGRES_PASSWORD": "*****",
+        "POSTGRES_HOST": POSTGRES_CONTAINER,
+        "POSTGRES_DB": POSTGRES_DB,
+    }
+    logger.info(f"Loaded Database configs {db_configs}")
```
Comment (AshciR) [inline src/shout_subgroup/database.py:25]: The logger wasn't setup correctly. Added these to help with that

File: src/shout_subgroup/database.py [hunk 11]
```diff
@@ -16,42 +16,28 @@ def configure_database() -> bool:
...
-    Base.metadata.drop_all(engine)
-    Base.metadata.create_all(engine)
-
     global session
 
     session_maker = sessionmaker(bind=engine)
     session = session_maker()
 
-    # Create users
-    richie = UserModel(telegram_user_id=12345, username="ashcir", first_name="Richie", last_name="Doe")
```
Comment (AshciR) [inline src/shout_subgroup/database.py:36]: No longer needed for application code

**Review by rick1e — APPROVED** [review]

**Pull request merged**